### ISSUE TITLE
Support filtering based on child table relationship

### ISSUE THREAD

**Issue by bossjapzz** [supabase/pg_graphql#88]
**Describe the bug**
Support filtering based on child table relationship

**Expected behavior**
<img width="395" alt="Screen Shot 2022-02-21 at 5 33 29 PM" src="https://user-images.githubusercontent.com/100037148/154927770-3ee428ee-a437-4f28-a637-484b1f68a207.png">

**Comment by apecollector** [comment]
Would like to see this as well, this is really what makes graphql appealing imo.

**Comment by gromchen** [comment]
How can I vote for this enhancement? 😅

**Comment by vwkd** [comment]
+1

[Hasura](https://hasura.io/docs/latest/queries/postgres/query-filters/#pg-nested-filter) can do this for both child and parent relationships. AFAIK the [REST API](https://supabase.com/docs/reference/javascript/using-filters) (see "Filter Foreign Tables") can do this currently only for child relationships.

For example with Hasura, this selects authors containing a name with all their books, or authors and the specific books whose description contains the name. Note how it filters both on child and parent relationships.

```graphql
query {
  authors(where: {_or: [{name: {_iregex: $name}}, {books: {description: {_iregex: $name}}}]}) {
    name
    books(where: {_or: [{author: {name: {_iregex: $name}}}, {description: {_iregex: $name}}]}) {
      title
      description
    }
  }
}
```

(Note, for the relationship names to be set like this in Hasura, the foreign key columns should *not* match the table name. For example, one can use a suffix like `_id`, e.g. `books_id` column in `authors` table and `author_id` column in `books` table.)

**Comment by robertn702** [comment]
Any update on this feature? It's currently my biggest blocker.

**Comment by cullophid** [comment]
I just wanted to add to the hype of this feature  🥳.

Keep up the good work pg_graphql is very promising!

**Comment by davidchalifoux** [comment]
I'm also looking for this feature!

**Comment by azlekov** [comment]
Interested in this one. Is there any workaround now like database views or whatever?

**Comment by olirice** [comment]
> Interested in this one. Is there any workaround now like database views or whatever?

unfortunately, there isn't currently a workaround for filtering parents based on children

**Comment by magrinj** [comment]
Hey @olirice,
Any update on this ?
We're using `pg_graphql` at [Twenty](https://github.com/twentyhq/twenty) when querying the database for the dynamic part of the GraphQL API, and really need this feature
Thanks a lot to Supabase for this awesome project !

**Comment by RobSchilderr** [comment]
> none yet, but we've identified nested inserts (which requires upsert support to work well) as the next big project we're tackling in pg_graphql so you should start to see movement on it

is there any workaround to achieve the same effect right now?

**Comment by olirice** [comment]
You could
- write a [function](https://supabase.github.io/pg_graphql/functions/) that performs the filter ahead of time
- use two requests / filter on the client
- use a [computed field](https://supabase.github.io/pg_graphql/computed_fields/) if you're always filtering the child in the same way (no parameters required)

but all have their drawbacks

**Comment by Mihai-github** [comment]
I'm trying to filter data from table A based on a value in table B. Specifically, I want to filter records where B.something some condition with a specific value. I haven't been able to find an example of this query in the Supabase documentation. Could you please confirm if this type of filtering is supported in GraphQL on Supabase? If not, are there any workarounds or alternative approaches I could consider?

**Comment by imor** [comment]
@Mihai-github have you tried one of the options mentioned by @olirice above? If these don't work for you can you share more details about your use case?

**Comment by imor** [comment]
> Is the GQL just translated into a pgREST query?

No, GQL is converted into a SQL query which is executed by the [`QueryEntrypoint::execute` method](https://github.com/supabase/pg_graphql/blob/b0032a5ad04eb583a3623e012a03615add68b4e4/src/transpile.rs#L78-L102) for a query and [`MutationEntrypoint::execute` method](https://github.com/supabase/pg_graphql/blob/b0032a5ad04eb583a3623e012a03615add68b4e4/src/transpile.rs#L44-L72) for a mutation.

> I'm not sure how the GQL is translated into a PG query

The `QueryEntrypoint::execute` and `MutationEntrypoint::execute` methods convert GQL into SQL by calling the [`self.to_sql_entrypoint` methods](https://github.com/supabase/pg_graphql/blob/b0032a5ad04eb583a3623e012a03615add68b4e4/src/transpile.rs#L76). The `to_sql_entrypoint` methods are implemented by various builder objects that implement the `QueryEntrypoint` or `MutationEntrypoint` traits. E.g. [`FunctionCallBuilder`'s impl](https://github.com/supabase/pg_graphql/blob/b0032a5ad04eb583a3623e012a03615add68b4e4/src/transpile.rs#L604-L608) just delegates to a call to its [`to_sql` method](https://github.com/supabase/pg_graphql/blob/b0032a5ad04eb583a3623e012a03615add68b4e4/src/transpile.rs#L552-L595). These builder objects are created earlier by parsing GQL and looking at the SQL context.

Let me know if this is sufficient detail for you to take a stab at implementing a solution or you need more.

**Comment by bh865** [comment]
We have any update on this?

> You could
> 
> * write a [function](https://supabase.github.io/pg_graphql/functions/) that performs the filter ahead of time
> * use two requests / filter on the client
> * use a [computed field](https://supabase.github.io/pg_graphql/computed_fields/) if you're always filtering the child in the same way (no parameters required)
> 
> but all have their drawbacks

Do you have a preferred method of these 3?

**Comment by olirice** [comment]
My preferred option would be #1
Extend with a function that performs the aggregation and then filter on that value
Docs: https://supabase.github.io/pg_graphql/computed_fields/#extending-types-with-functions

**Comment by bh865** [comment]
> My preferred option would be #1 Extend with a function that performs the aggregation and then filter on that value Docs: https://supabase.github.io/pg_graphql/computed_fields/#extending-types-with-functions

So, I've managed to get a function that I can query successfully in the SQL Editor but it does not show up in my GraphiQL playground. Any suggestions? I left out 'immutable' and 'stable' as I've tried both to no avail.

`CREATE OR REPLACE FUNCTION filter_events_search(gender VARCHAR)
  RETURNS TABLE(
    id UUID
  )
LANGUAGE plpgsql
AS $$
BEGIN
  RETURN QUERY
  SELECT
    e.id
  FROM 
    events e 
    JOIN teams t ON e.team_id = t.id
  WHERE
    t.gender = filter_events_search.gender;
END;
$$;`

**Comment by bobbybol** [comment]
Hi @olirice, any progress on this? Thanks ;)

**Comment by olirice** [comment]
not yet. we're currently working on upsert

**Comment by phenric** [comment]
Any update? Is the development planned?

**Comment by olirice** [comment]
hasn't been planned. this is a tricky one both in interface and writing a generic resolver

**Comment by maxcan** [comment]
Its tragic that we don't have this.  Hugely useful feature from Hasura. @vwkd's example is something i used all the time

**Comment by derekjwilliams** [comment]
> Its tragic that we don't have this. Hugely useful feature from Hasura. [@vwkd](https://github.com/vwkd)'s example is something i used all the time. [@olirice](https://github.com/olirice) the Hasura GraphQL Engine (v2) is open source. Can take inspiration from that perhaps?

It may also be worth taking a look a the Postgraphile (https://github.com/graphile/crystal) implementation for inspiration

**Comment by gio-toptal** [comment]
This is really sad, 2026 and still no plans for this?
